src/linux_android.rs

```diff
@@ -10,77 +10,90 @@
 extern crate libc;
 
 use Error;
-use utils::use_init;
 use std::fs::File;
-use std::io;
+use std::{io, mem, thread};
 use std::io::Read;
-use std::cell::RefCell;
 use std::num::NonZeroU32;
-use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::os::unix::io::{AsRawFd, FromRawFd};
 
-static RNG_INIT: AtomicBool = AtomicBool::new(false);
+// replace with AtomicU8 on stabilization and MSRV bump
+static RNG_STATE: AtomicUsize = AtomicUsize::new(0);
+// replace with AtomicI32 on stabilization and MSRV bump
+static RNG_FD: AtomicUsize = AtomicUsize::new(0);
 
-enum RngSource {
-    GetRandom,
-    Device(File),
+const STATE_USE_SYSCALL: usize = 1 << 0;
+const STATE_USE_FD: usize = 1 << 1;
+const STATE_INIT: usize = 1 << 2;
+
+pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
+    let state = RNG_STATE.load(Ordering::Relaxed);
+    if state & STATE_USE_SYSCALL != 0 {
+        use_syscall(dest)
+    } else if state & STATE_USE_FD != 0 {
+        use_fd(dest)
+    } else {
+        init_loop(dest)
+    }
 }
 
-thread_local!(
-    static RNG_SOURCE: RefCell<Option<RngSource>> = RefCell::new(None);
-);
+fn init_loop(dest: &mut [u8]) -> Result<(), Error> {
+    loop {
+        let mut state = RNG_STATE.fetch_or(STATE_INIT, Ordering::AcqRel);
+
+        if state & STATE_INIT != 0 {
+            thread::yield_now();
+            continue;
+        }
+        if state & (STATE_USE_SYSCALL | STATE_USE_FD) == 0 {
+            state = init()?;
+        }
+        if state & STATE_USE_SYSCALL != 0 {
+            return use_syscall(dest);
+        } else if state & STATE_USE_FD != 0 {
+            return use_fd(dest);
+        }
+    }
+}
+
+fn init() -> Result<usize, Error> {
+    match use_syscall(&mut [0u8; 0]) {
+        Ok(()) => {
+            RNG_STATE.store(STATE_USE_SYSCALL, Ordering::Release);
+            Ok(STATE_USE_SYSCALL)
+        },
+        Err(err) if err.code().get() as i32 == libc::ENOSYS => {
+            File::open("/dev/random")?.read_exact(&mut [0u8; 1])?;
+            let fd = File::open("/dev/urandom")?.as_raw_fd();
+            // it's safe to convert RawFd to usize and back because nonnegativity
+            // is guaranteed (see `open` man)
+            RNG_FD.store(fd as usize, Ordering::Release);
+            RNG_STATE.store(STATE_USE_FD, Ordering::Release);
+            Ok(STATE_USE_FD)
+        },
+        Err(err) => Err(err),
+    }
+}
 
-fn syscall_getrandom(dest: &mut [u8]) -> Result<(), io::Error> {
+fn use_syscall(dest: &mut [u8]) -> Result<(), Error> {
     let ret = unsafe {
         libc::syscall(libc::SYS_getrandom, dest.as_mut_ptr(), dest.len(), 0)
     };
     if ret < 0 || (ret as usize) != dest.len() {
-        error!("Linux getrandom syscall failed with return value {}", ret);
-        return Err(io::Error::last_os_error());
+        // replace with errno?
+        return Err(io::Error::last_os_error().into());
```

Comment:
I don't think `errno` is exposed in either `std::os` or `libc`? Anyway, this method of capturing it is sufficient.

According to the `errno` man page, the value must be saved immediately after the syscall, so the old

Reply:
No, it's not, but it's possible to copy a bit of code from [here](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/os.rs#L29). The problem with `io::Error::last_os_error().into()` is that IIRC it does not get optimized into pure `errno` call, so there is a small overhead.